pkg/counter/cache.go

```diff
@@ -0,0 +1,162 @@
+	operation()
+}
+
+// Reset resets the cached counter - is supposed to be used only in tests
+func Reset() {
+	write(func() {
+		reset()
+	})
+}
+
+func reset() {
+	cachedCounts.Counts = Counts{
+		UserAccountsPerClusterCounts: map[string]int{},
+	}
+	cachedCounts.initialized = false
+}
+
+// IncrementMasterUserRecordCount increments the number of MasterUserRecord in the cached counter
+func IncrementMasterUserRecordCount() {
+	write(func() {
+		cachedCounts.MasterUserRecordCount++
+	})
+}
+
+// DecrementMasterUserRecordCount decreases the number of MasterUserRecord in the cached counter
+func DecrementMasterUserRecordCount(log logr.Logger) {
+	write(func() {
+		if cachedCounts.MasterUserRecordCount != 0 {
+			cachedCounts.MasterUserRecordCount--
+		} else {
+			log.Error(fmt.Errorf("the count of MasterUserRecords is zero"),
+				"unable to decrement the number of MasterUserRecords")
+		}
+	})
+}
+
+// IncrementUserAccountCount increments the number of UserAccount fo the given member cluster in the cached counter
+func IncrementUserAccountCount(clusterName string) {
+	write(func() {
+		cachedCounts.UserAccountsPerClusterCounts[clusterName]++
+	})
+}
+
+// DecrementUserAccountCount decreases the number of UserAccount fo the given member cluster in the cached counter
+func DecrementUserAccountCount(log logr.Logger, clusterName string) {
+	write(func() {
+		if cachedCounts.UserAccountsPerClusterCounts[clusterName] != 0 {
+			cachedCounts.UserAccountsPerClusterCounts[clusterName]--
+		} else {
+			log.Error(fmt.Errorf("the count of UserAccounts is zero"),
+				"unable to decrement the number of UserAccounts for the given cluster", "cluster", clusterName)
+		}
+	})
+}
+
+// GetCounts returns Counts struct containing number of MURs and number of UserAccounts per member cluster.
+// If the counter is not yet initialized, then it returns error
+func GetCounts() (Counts, error) {
+	cachedCounts.RLock()
+	defer cachedCounts.RUnlock()
+	if !cachedCounts.initialized {
+		return cachedCounts.Counts, fmt.Errorf("counter is not initialized")
+	}
+	return cachedCounts.Counts, nil
+}
+
+// Synchronize synchronizes the content of the ToolchainStatus with the cached counter
+//
+// If the counter hasn't been initialized yet, then it adds the actual cached count
+// to the one taken from ToolchainStatus and marks the cached as initialized
+//
+// If the ToolchainStatus doesn't contain any MUR counts, then it lists all existing
+// MURs, counts them and stores in both cache and ToolchainStatus object
+//
+// If the cached counter is initialized and ToolchainStatus contains already some numbers
+// then it updates the ToolchainStatus numbers with the one taken from the cached counter
+func Synchronize(cl client.Client, toolchainStatus *v1alpha1.ToolchainStatus) error {
+	cachedCounts.Lock()
+	defer cachedCounts.Unlock()
+	if toolchainStatus.Status.HostOperator == nil || toolchainStatus.Status.HostOperator.CapacityUsage.MasterUserRecordCount == 0 {
```

Comment:
why do you need to check if `toolchainStatus.Status.HostOperator.CapacityUsage.MasterUserRecordCount == 0`? Isn't `toolchainStatus.Status.HostOperator==nil` enough? (and we could have 0 MUR when the platform is just started)

Reply:
It's not enough, the `toolchainStatus.Status.HostOperator` will be set by `ToolchainStatus` controller  - I'm checking if it's nil just for my own paranoia - I don't like such panics :grimacing: 
Anyway, I need to check the case when this feature is deployed to the cluster for the first time - in that case the count is zero and I need to list all present MURs and count them.